src/estimation/src/AttitudeQuaternionEKF.cpp

```diff
@@ -0,0 +1,594 @@
+      m_state.m_orientation(0) = 1.0;
+      m_state.m_angular_velocity.zero();
+      m_state.m_gyroscope_bias.zero();
+      m_initial_state = m_state;
+
+      m_orientationInSO3.fromQuaternion(m_state.m_orientation);
+      m_orientationInRPY = m_orientationInSO3.asRPY();
+
+      m_Omega_y.zero();
+      m_Acc_y.zero();
+
+      m_gravity_direction.zero();
+      m_gravity_direction(2) = 1;
+
+      iDynTree::toEigen(m_Id4).setIdentity();
+      iDynTree::toEigen(m_Id3).setIdentity();
+}
+
+bool iDynTree::AttitudeQuaternionEKF::initializeFilter()
+{
+    m_state_size = this->getInternalStateSize();
+
+    if (m_params.use_magenetometer_measurements)
+    {
+        m_output_size = output_dimensions_with_magnetometer;
+    }
+    else
+    {
+        m_output_size = output_dimensions_without_magnetometer;
+    }
+
+    ekfSetStateSize(m_state_size);
+    m_x.resize(m_state_size);
+    m_w.resize(m_state_size);
+    serializeStateVector();
+    ekfSetOutputSize(m_output_size);
+
+    m_y.resize(m_output_size);
+    m_v.resize(m_output_size);
+    m_input_size = input_dimensions;
+    ekfSetInputSize(m_input_size);
+    m_u.resize(m_input_size);
+
+    // ekfinit resizes and initializes the filter matrices and vectors to zero - dont move it frome here
+    if (!ekfInit())
+    {
+        return false;
+    }
+
+    // once the buffers are resized and zeroed, setup the matrices
+    if (!setInitialStateCovariance(m_params.initial_orientation_error_variance,
+                                   m_params.initial_ang_vel_error_variance,
+                                   m_params.initial_gyro_bias_error_variance))
+    {
+        return false;
+    }
+
+    if (!setSystemNoiseVariance(m_params.gyroscope_noise_variance,
+                            m_params.gyro_bias_noise_variance))
+    {
+        return false;
+    }
+
+    if (!setMeasurementNoiseVariance(m_params.accelerometer_noise_variance,
+                                m_params.magnetometer_noise_variance))
+    {
+        return false;
+    }
+
+    m_initialized = true;
+    return m_initialized;
+}
+
+bool iDynTree::AttitudeQuaternionEKF::propagateStates()
+{
+    iDynTree::Span<double> Omega_y_span(m_Omega_y.data(), m_Omega_y.size());
+    bool ok  = ekfSetInputVector(Omega_y_span);
+    ok = ekfPredict() && ok;
+    iDynTree::Span<double> x_span(m_x.data(), m_x.size());
+    if (ekfGetStates(x_span))
+    {
+        deserializeStateVector();
+        m_orientationInSO3 = iDynTree::Rotation::RotationFromQuaternion(m_state.m_orientation);
+        m_orientationInRPY = m_orientationInSO3.asRPY();
+    }
+    else
+    {
+        iDynTree::reportError("AttitudeQuaternionEKF", "updateFilterWithMeasurements", "could not get recent state estimate");
+        return false;
+    }
+    return ok;
+}
+
+void iDynTree::AttitudeQuaternionEKF::serializeMeasurementVector()
+{
+    if (m_y.size() != m_output_size)
+    {
+        m_y.resize(m_output_size);
+    }
```

Comment:
If I am not wrong, `VectorDynSize` already contains the size check inside, so I guess you can just resize the vector with no additional check.

Reply:
I had the intention of resizing the buffers only when m_output_size has changed due to some RPC calls. In that perspective, checking the size and deciding to resize is a less complex operation  than resizing it every time. 
Would you agree?